File: Python/NumberOfWaysToAssignEdgeWeights_I.py
class Solution(object):
    MOD = 1000000007

    def assignEdgeWeights(self, edges):
        """
        :type edges: List[List[int]]
        :rtype: int
        """
        n = len(edges) + 1

        graph = [[] for _ in range(n + 1)]
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)

        depth = [-1] * (n + 1)
        depth[1] = 0

        q = [1]
        front = 0
        max_depth = 0

        while front < len(q):
            u = q[front]
            front += 1

            max_depth = max(max_depth, depth[u])

            for v in graph[u]:
                if depth[v] == -1:
                    depth[v] = depth[u] + 1
                    q.append(v)

        if max_depth == 0:
            return 0

        return pow(2, max_depth - 1, self.MOD)

File: Java/NumberOfWaysToAssignEdgeWeights_I.java
import java.util.*;

class NWAEW_I {
    private static final long MOD = 1_000_000_007L;

    public int assignEdgeWeights(int[][] edges) {
        int n = edges.length + 1;

        List<Integer>[] graph = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int[] e : edges) {
            int u = e[0];
            int v = e[1];
            graph[u].add(v);
            graph[v].add(u);
        }

        Queue<Integer> q = new LinkedList<>();
        int[] depth = new int[n + 1];
        Arrays.fill(depth, -1);

        q.offer(1);
        depth[1] = 0;

        int maxDepth = 0;

        while (!q.isEmpty()) {
            int u = q.poll();
            maxDepth = Math.max(maxDepth, depth[u]);

            for (int v : graph[u]) {
                if (depth[v] == -1) {
                    depth[v] = depth[u] + 1;
                    q.offer(v);
                }
            }
        }

        if (maxDepth == 0) return 0;

        return (int) modPow(2, maxDepth - 1);
    }

    private long modPow(long base, long exp) {
        long res = 1;

        while (exp > 0) {
            if ((exp & 1) == 1) {
                res = (res * base) % MOD;
            }
            base = (base * base) % MOD;
            exp >>= 1;
        }

        return res;
    }
}

public class NumberOfWaysToAssignEdgeWeights_I {
    
}
